rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/DisconnectionRouter.java

```diff
@@ -0,0 +1,74 @@
+package com.polidea.rxandroidble.internal.connection;
+
+
+import com.jakewharton.rxrelay.BehaviorRelay;
+import com.polidea.rxandroidble.RxBleAdapterStateObservable;
+import com.polidea.rxandroidble.exceptions.BleDisconnectedException;
+import com.polidea.rxandroidble.exceptions.BleException;
+import com.polidea.rxandroidble.internal.DeviceModule;
+import javax.inject.Inject;
+import javax.inject.Named;
+import rx.Observable;
+import rx.functions.Func1;
+
+/**
+ * A class that is responsible for routing all potential sources of disconnection to an Observable that emits only errors.
+ */
+@ConnectionScope
+class DisconnectionRouter {
+
+    private final BehaviorRelay<BleException> disconnectionErrorRelay = BehaviorRelay.create();
+
+    private final Observable disconnectionErrorObservable;
+
+    @Inject
+    DisconnectionRouter(
+            @Named(DeviceModule.MAC_ADDRESS) final String macAddress,
+            Observable<RxBleAdapterStateObservable.BleAdapterState> adapterStateObservable
+    ) {
+        disconnectionErrorObservable = Observable.merge(
+                disconnectionErrorRelay
+                        .flatMap(new Func1<BleException, Observable<?>>() {
+                            @Override
+                            public Observable<?> call(BleException e) {
+                                return Observable.error(e);
+                            }
+                        }),
+                adapterStateObservable
+                        .filter(new Func1<RxBleAdapterStateObservable.BleAdapterState, Boolean>() {
+                            @Override
+                            public Boolean call(RxBleAdapterStateObservable.BleAdapterState bleAdapterState) {
+                                return !bleAdapterState.isUsable();
+                            }
+                        })
+                        .flatMap(new Func1<RxBleAdapterStateObservable.BleAdapterState, Observable<?>>() {
+                            @Override
+                            public Observable<?> call(RxBleAdapterStateObservable.BleAdapterState bleAdapterState) {
+                                return Observable.error(new BleDisconnectedException(macAddress)); // TODO: Introduce BleDisabledException?
+                            }
+                        })
+        )
+                .replay()
```

Comment:
Do we need replay & autoConnect at the observable here or just when listening for adapter state?

Reply:
In my opinion yes as we would like to have it repeated to anyone who will subscribe later

Reply:
Isn't a `BehaviorRelay` enough at the source?

Reply:
Actually `disconnectionErrorRelay` should be a `PublishRelay`.
With this `.replay().autoConnect(0)` we are sure that both the `adapterStateObservable()` and `disconnectionErrorRelay` will get unsubscribed when any error/emission will happen

Reply:
Ok, agreed.